raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -129,13 +135,17 @@
     private final ConnectionCache connections;
     private final FuturePurgatory<Void> purgatory;
     private final BlockingQueue<PendingAppendRequest> unsentAppends;
+    private final SortedMap<OffsetAndEpoch, BaseOffsetAndTime> pendingAppends;
```

Comment:
The replication metric needs to remember which appended records gets eventually committed. Personally I think this is what most state-machine impls have to do anyways, so it's a bit duplicated code (see SimpleKeyValueStore has a similar logic as well).